devito/types/basic.py

```diff
@@ -1439,6 +1447,30 @@ def _offset_owned(self):
 
         return DimensionTuple(*offsets, getters=self.dimensions, left=left, right=right)
 
+    @cached_property
+    def _offset_subdomain(self):
+        """Offset of subdomain indices versus the global index."""
+        # If defined on a SubDomain, then need to offset indices accordingly
+        if not self._is_on_subdomain:
+            return (0,)*len(self.dimensions)
+        # Symbolic offsets to avoid potential issues with user overrides
+        offsets = []
+        for d in self.dimensions:
+            if d.is_Sub:
+                l_tkn, r_tkn = d.tkns
+                if l_tkn.value is None:
+                    # Right subdimension
+                    offsets.append(-r_tkn + d.symbolic_max + 1)
+                elif r_tkn.value is None:
+                    # Left subdimension
+                    offsets.append(0)
+                else:
+                    # Middle subdimension
+                    offsets.append(l_tkn)
+            else:
+                offsets.append(0)
+        return tuple(offsets)
```

Comment:
DimensionTuple?

Reply:
Don't think so

Reply:
I second Mathias here, I think DimensionTuple(...) makes much more sense, for homogeneity in particular